### PR TITLE
Implement Disover::Service#each_update

### PR DESCRIPTION
**Pull request by lmars**
@titanous interested to get your thoughts on the approach here, I have always used EventMachine in the past so this is my first project using Celluloid

### CODE REVIEW

**Comment by titanous** [comment]
Looks like it needs a rebase to get Travis going.

**Comment by lmars** [comment]
@titanous yup rebased

File: lib/discover.rb [hunk 1]
```diff
@@ -106,12 +106,36 @@ def sentinel?
...
+
+      def initialize(block)
+        @block   = block
+        @updates = []
+        @future  = Celluloid::Future.new
+      end
+
+      def notify(update)
+        @block.call update
+        @updates << update
```
Comment (titanous) [inline lib/discover.rb]: What is the reason why you're storing all of the updates?

File: lib/discover.rb [hunk 2]
```diff
@@ -126,7 +150,10 @@ def offline
     end
 
     def each_update(&block)
```
Comment (titanous) [inline lib/discover.rb]: Add a flag that specifies whether to send all of the online services first, or just new updates.

**Comment by titanous** [comment]
Sorry for the spam, I commented on the wrong tab.

File: lib/discover.rb (hunk 1 above)
Comment (lmars) [inline lib/discover.rb]: Ah I did a refactor and must of left this in. Originally I wanted to use `each_update.to_a` in the tests (hence why I signal the future with the updates), but that didn't really make sense.

I've removed this and just used a condition variable: 46faa78

File: lib/discover.rb (hunk 2 above)
Comment (lmars) [inline lib/discover.rb]: Added: ec3d694

File: lib/discover.rb [hunk 3]
```diff
@@ -106,12 +106,34 @@ def sentinel?
...
+      def initialize(block)
+        @block     = block
+        @condition = Condition.new
+      end
+
+      def notify(update)
+        @block.call update
+      end
+
+      def finish
```
Comment (titanous) [inline lib/discover.rb]: What about calling this `done`?

**Comment by titanous** [comment]
LGTM.

File: lib/discover.rb (hunk 3 above)
Comment (lmars) [inline lib/discover.rb]: yup updated: https://github.com/flynn/discoverd-ruby/blob/service-each-update/lib/discover.rb#L121

**Comment by lmars** [comment]
@titanous cleaned up and ready to merge

**Pull request merged**